queue_job/models/queue_job.py

```diff
@@ -235,6 +238,70 @@ def requeue(self):
         self._change_job_state(PENDING)
         return True
 
+    def button_terminate(self):
+        self.terminate()
+        return {
+            "type": "ir.actions.client",
+            "tag": "reload",
+        }
+
+    def terminate(self, raise_error=True):
+        """Attempt to kill the backend process related to the running job."""
+        error_jobs = self.env["queue.job"]
+        for record in self:
+            # Check if the job can be terminated
+            if record.state != STARTED:
+                error_message = _(
+                    f"Unable to terminate job {record.uuid} because it is not started"
+                )
+                _logger.warning(error_message)
+                if raise_error:
+                    raise exceptions.UserError(error_message)
+                continue
+
+            elif not record.db_backend_pid:
+                error_message = _(
+                    f"Unable to terminate job {record.uuid}. Missing 'db_backend_pid'",
+                )
+                _logger.warning(error_message)
+                if raise_error:
+                    raise exceptions.UserError(error_message)
+                continue
+
+            # Compare xact_start and job date_started as reference
+            query_vals = {"pid": record.db_backend_pid}
+            self.env.cr.execute(
+                "SELECT xact_start FROM pg_stat_activity WHERE pid = %(pid)s;",
+                query_vals,
+            )
+            backend_start = self.env.cr.fetchone()
+            if backend_start:
+
+                job_start = pytz.utc.localize(record.date_started)
+                difference = abs(job_start - backend_start[0])
+
+                if difference < timedelta(seconds=1):
```

Comment:
One second is a long time for a database transaction. I see that in date_started, the microseconds are not stored, and its value is not determined by the SQL transaction so I understand this is the best check possible at this point. However, as a sanity mechanism it is too dangerous. I don't think this way we can guarantee not to shoot down backends that have moved on from the job transaction.

There is an alternative, and that is to use store and use the transaction id from the cursor (in addition to, or even instead of storing the backend pid):
```
essentpg11=> select backend_xid from pg_stat_activity where state != 'idle';
 backend_xid 
-------------
      414063
(1 row)

essentpg11=> select txid_current()                                    
;
 txid_current 
--------------
       414063
(1 row)
```

Reply:
I refactored the PR to use `pg_current_xact_id()` or `txid_current()` and then use `pg_stat_activity` to get the right backend PID as you suggested @StefanRijnhart.
Notes:
- I changed the field type for `db_txid` to from Integer to Char because bigints are not yet supported by the Odoo core in V14. Would you agree that a textual representation is acceptable for now?
- The termination method now casts the text string back to an xid which is a valid selector for the query.